### ISSUE TITLE
staticcheck: detect unintentional encoder and decoder implementations

### ISSUE THREAD

**Issue by ainar-g** [dominikh/go-tools#715]
<p>
(Couldn't think of a more descriptive title, feel free to change.)
</p>

<p>
See
<a href="https://twitter.com/katie_hockman/status/1242527795777069062">this tweet</a>
from @katiehockman.
</p>

``` go
func main() {
	t := struct {
		time.Time
		N int
	}{time.Now(), 5}

	m, _ := json.Marshal(t)
	fmt.Printf("%s", m)
}
```

<p>
This will print <code>2009-11-10T23:00:00Z</code>, because
<code>time.Time</code> implements <code>json.Marshaler</code>.
</p>

<p>
The structs that we can definitely mark are the structs that (1) have
fields with <code>json</code> tags, (2) embed marshalers, and (3) don't
implement the interface themselves:
</p>

``` go
type bad struct {
	time.Time
	N int `json:"n"`
}
```

**Comment by FiloSottile** [comment]
I'd argue that any type that embeds a type that implements TextMarshaler, BinaryMarshaler, or json.Marshaler (and unmarshalers) should get flagged. It's almost certainly not intended, and if it is it can be implemented in the parent type by calling the embedded method.

**Comment by ainar-g** [comment]
<p>
@FiloSottile  I think, that's way too radical.  Consider something like:
</p>

``` go
type TimeWithSomeContext struct {
	time.Time

	Processed bool
	Status    int
	// …
}
```

<p>
I want <code>TimeWithSomeContext</code> to look the same as the usual
time in my JSON, because the users of the JSON API don't care about
<code>Processed</code>, <code>Status</code>, etc., so embedding
<code>time.Time</code> is perfectly fine here.  Forcing all programmers
to add a <code>MarshalJSON</code> method here would be too much, and
also not quite something that's in staticcheck's philosophy (at least as
I understand it from Dominik's previous comments).
</p>

**Comment by broady** [comment]
@ainar-g you'd do this instead:

```
type TimeWithSomeContext struct {
	Time time.Time

	Processed bool
	Status    int
	// …
}

func (t TimeWithSomeContext) MarshalJSON() ([]byte, error) {
  return t.Time.MarshalJSON()
}
```

**Comment by dominikh** [comment]
I agree with Ainar. We can very likely flag structs that use tags, but flagging any struct that embeds a marshaler will lead to some false positives, and we always try to err on the side of false negatives, not false positives. It's fine to miss some bugs, it's not fine to claim bugs where there aren't any.

We _could_ consider adding an opt-in for more strict checking a la Filippo, and then you can decide to enforce that rule in your own code base.

**Comment by FiloSottile** [comment]
Could you also flag locations where such a type is passed to json.Marshal?

**Comment by dominikh** [comment]
> Could you also flag locations where such a type is passed to json.Marshal?
The statically known ones, yes. For the case of embedding but no struct tags, we can _only_ flag the locations where `json.Marshal` etc are being called; merely embedding a marshaler in itself is not an error if the type is never being marshaled.
